### PR TITLE
[CHORE] Centralize pyo3 pickling around `__reduce__` + bincode macro.

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR ports all pyo3-exposed classes that need to be picklable to use a `__reduce__` + bincode macro, which removes the need for a dummy/inefficient no-arg constructor.

### CODE REVIEW

File: src/daft-core/src/datatypes/image_format.rs [hunk 1]
```diff
@@ -11,7 +11,7 @@ use common_error::{DaftError, DaftResult};
 /// Supported image formats for Daft's I/O layer.
 #[allow(clippy::upper_case_acronyms)]
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, Hash)]
-#[cfg_attr(feature = "python", pyclass)]
+#[cfg_attr(feature = "python", pyclass(module = "daft.daft"))]
```
Comment (samster25) [inline src/daft-core/src/datatypes/image_format.rs:14]: could we make an attr that is `daftpyclass=pyclass(module = "daft.daft")` and use that everywhere?

File: src/daft-core/src/python/datatype.rs [hunk 2]
```diff
@@ -283,49 +273,51 @@ impl PyDataType {
...
-                    .to_object(py)),
-                (_, _) => ffi::to_py_schema(&self.dtype.to_arrow()?, py, pyarrow)?
-                    .getattr(py, pyo3::intern!(py, "type")),
-            }
-        })
+    pub fn to_arrow(
+        &self,
+        py: Python,
+        cast_tensor_type_for_ray: Option<bool>,
+    ) -> PyResult<PyObject> {
```
Comment (samster25) [inline src/daft-core/src/python/datatype.rs:280]: is this a bad diff? what's different

File: src/daft-core/src/python/field.rs [hunk 3]
```diff
@@ -38,22 +24,10 @@ impl PyField {
...
-            Err(e) => Err(e),
-        }
-    }
-
-    pub fn __getstate__(&self, py: Python) -> PyResult<PyObject> {
-        Ok(PyBytes::new(py, &bincode::serialize(&self.field).unwrap()).to_object(py))
-    }
 }
 
+impl_bincode_py_state_serialization!(PyField);
```
Comment (samster25) [inline src/daft-core/src/python/field.rs:29]: this might be a good macro to implement as a derive so we can `#[derive(PySerDe)]`

**Review by samster25 — COMMENTED** [review]
Very nice! Thanks for taking this on

File: src/daft-core/src/python/datatype.rs (hunk 2 above)
Comment (clarkzinzow) [reply src/daft-core/src/python/datatype.rs:280]: Ah, this is a driveby! This was me noticing that we were redundantly capturing the GIL with `Python::with_gil(|py| ...)` when pyo3 methods already have the GIL, and the `py: Python` marker token can be transparently injected into the method by specifying it as an argument.

**Pull request merged**